### PR TITLE
Fixed deadlock in System.Resources.ResourceSet

### PR DESCRIPTION
**Pull request by PashaPash**
Fixed deadlock between ResourceSet iterator and and ResourceSet.GetObject.

Using both ResourceSet iterator and  ResourceSet.GetObject is a common scenario in web app - resource set is used for server view rendering (through ResourceSet.GetObject) and in the time that resource set is offloaded to client side as JSON (through ResourceSet iterator). There is a high probability that app will deadlock on start in that case.

Connect Issue: [#580259](https://connect.microsoft.com/VisualStudio/feedback/details/580259/deadlocking-under-load-in-system-resources-resourceset) - closed as non-reproducable. I have no access to the source code in that connect issue, but deadlock graph seems to be exactly the same as in my sample app.

Sample app to reproduce the issue: https://github.com/PashaPash/ResManDeadlock

Details from [Stack Exchange thread](http://stackoverflow.com/questions/17965347/foreach-mysteriously-hangs-on-first-item-of-a-resourceset):

The problem is that when using a ResourceSet the iterator takes out locks on the internal resource cache of the ResourceReader http://referencesource.microsoft.com/#mscorlib/system/resources/resourcereader.cs,1389 and then in the method AllocateStringNameForIndex takes out a lock on the reader itself: http://referencesource.microsoft.com/#mscorlib/system/resources/resourcereader.cs,447

    lock (_reader._resCache) {
         key = _reader.AllocateStringForNameIndex(_currentName, out _dataPosition); // locks the reader

Getting an object takes out the same locks int the opposite order: 

http://referencesource.microsoft.com/#mscorlib/system/resources/runtimeresourceset.cs,300
and http://referencesource.microsoft.com/#mscorlib/system/resources/runtimeresourceset.cs,335 

    lock(Reader) {
        ....
        lock(_resCache) {
           _resCache[key] = resLocation;
        }
    }

### CODE REVIEW

**Comment by PashaPash** [comment]
I updated the commit message to match contribution guidelines

**Comment by AlexGhiondea** [comment]
Thanks for the contribution. At first glance the change look ok.

However, since this code is shared with Desktop we need to run this through additional validation to make sure we don't break anything.

Once we are sure we are not breaking anything we are going to merge it in.

@ramarag can you take a look at this?

File: src/mscorlib/src/System/Resources/ResourceReader.cs [hunk 1]
```diff
@@ -1373,8 +1373,9 @@ public bool MoveNext()
 
                     String key;
                     Object value = null;
+                    lock (_reader) { // locks should be taken in the same order as in RuntimeResourceSet.GetObject to avoid deadlock
                     lock (_reader._resCache) {
```
Comment (ramarag) [inline src/mscorlib/src/System/Resources/ResourceReader.cs]: Can you fix the indentation here. I am almost done running the tests. will merge when I get the updated PR

Comment (PashaPash) [inline src/mscorlib/src/System/Resources/ResourceReader.cs]: @ramarag, sorry for broken indentation, just tried to keep number of lines changed at minimum.
I added an indentation and rebased new commit to master). Please review.

**Comment by PashaPash** [comment]
I added an indentation and rebased commit to master. Please review.

**Comment by ramarag** [comment]
LGTM... @jkotas  can you take a look too.

**Comment by jkotas** [comment]
LGTM

**Comment by jkotas** [comment]
Thank you for getting this fixed!

**Pull request merged**